src/lsm/grid.zig

```diff
@@ -768,33 +707,102 @@ pub fn GridType(comptime Storage: type) type {
                 grid.read_resolving = false;
             }
 
-            if (block) |b| {
-                const header = mem.bytesAsValue(vsr.Header, b[0..@sizeOf(vsr.Header)]);
-                assert(header.op == read.address);
-                assert(header.checksum == read.checksum);
-                maybe(read.repair);
-            } else {
-                assert(read.repair);
-            }
+            const header = mem.bytesAsValue(vsr.Header, block[0..@sizeOf(vsr.Header)]);
+            assert(header.op == read.address);
+            assert(header.checksum == read.checksum);
 
             // Resolve all reads queued to the address with the block.
             while (read.resolves.pop()) |pending| {
                 const pending_read = @fieldParentPtr(Read, "pending", pending);
                 assert(pending_read.address == read.address);
                 assert(pending_read.checksum == read.checksum);
-                assert(pending_read.repair == read.repair);
 
-                switch (pending_read.callback) {
-                    .normal => |callback| callback(pending_read, block.?),
-                    .repair => |callback| callback(pending_read, block),
-                }
+                pending_read.callback(pending_read, block);
             }
 
             // Then invoke the callback with the cache block (which should be valid for the duration
             // of the callback as any nested Grid calls cannot synchronously update the cache).
-            switch (read.callback) {
-                .normal => |callback| callback(read, block.?),
-                .repair => |callback| callback(read, block),
+            read.callback(read, block);
+        }
+
+        /// If the block is not present (or corrupt), we do not attempt to repair it — the repair's
+        /// address/checksum is not assumed to match what "should" be in our block.
+        /// Relatedly we still attempt to read free blocks — they might still hold the requested data.
```

Comment:
Is there a semantic difference between "not present" and "free" (does the former refer to the cache)?

Reply:
"Not present" means that we don't have the block in our cache _or_ disk.
"Free" means that the block's address is not held by the freeset.